Fb 360 calendar

ukmo-juan-castillo opened the pull request:
Ready to be reviewed, se ticket 209:

https://github.com/NOAA-EMC/WW3/issues/209

ukmo-juan-castillo committed:
Changes to add support to 360-day calendar - see ticket #209
https://github.com/NOAA-EMC/WW3/issues/209
At this point, only code changes are added. Code functionality is
already tested, regtests not run yet.

ukmo-juan-castillo committed:
Changes to add support to 360-day calendar - see ticket #209
https://github.com/NOAA-EMC/WW3/issues/209
At this point, only code changes are added. Code functionality is
already tested, regtests not run yet.

ukmo-juan-castillo committed:
Minor fixes and changes for adding and using regtest

ukmo-juan-castillo committed:
Minor fixes and changes for adding and using regtest

ukmo-juan-castillo committed:
Move the 360-day calendar flag to the MISC namelist

ukmo-juan-castillo committed:
Move the 360-day calendar flag to the MISC namelist

ukmo-juan-castillo committed:
Merge branch 'develop' into fb_360_calendar

ukmo-juan-castillo committed:
Latest changes before review: update version number of latest changes

ukmo-juan-castillo committed:
Latest changes before review: update version number of latest changes

ukmo-juan-castillo committed:
Revert empty spaces that were inadvertingly introduced

ukmo-juan-castillo committed:
Revert empty spaces that were inadvertingly introduced

ukmo-ccbunney commented:
Will also add @mickaelaccensi for external review (collaborator status pending)

ukmo-ansaulter commented on model/ftn/ww3_grid.ftn:
```diff
@@ -2582,6 +2585,15 @@
       WRITE (NDSO,967) FHMAX, FHMAX/SQRT(2.), YESXNO(J)
       IF ( FHMAX.LT.0.50 .AND. J.EQ.1 ) WRITE (NDST,968)
+!
+      IF ( CAL360 ) THEN
```
Might it be better to deploy the catch for when CAL360 and NOLEAP are set true, used at line 6293 here? As written I think the code would only deal with this problem if reading from the .nml file?

ukmo-ansaulter replied:
Thanks Juan, the new changes using CALTYPE look good here

ukmo-ansaulter commented on model/ftn/ww3_grid.ftn:
```diff
@@ -6276,6 +6290,7 @@
                           READ (NDS,NML=OUTS,END=801,ERR=802,IOSTAT=J)
                         CASE('MISC')
                           READ (NDS,NML=MISC,END=801,ERR=802,IOSTAT=J)
+                          IF ( CAL360 .AND. NOLEAP ) GOTO 805
```
See comment at line 2589

ukmo-juan-castillo replied:
Agree with the change of location for the check on CAL360 and NOLEAP. Waiting for suggestions on where to add the 360-day calendar regtest.

When using namelist files, the calendar is read from the file specified by the parameter NML_GRID%NML. In the template file this is pointing to the file 'namelist.nml', which does not exist.

mickaelaccensi replied:
namelist.nml exists for each regtest

ukmo-juan-castillo replied:
What I meant is that the namelist.nml template file does not exist. But of course, in the new regtest to check the 360-day calendar the CAL360 parameter is properly set. In the rest of the regtests it is not included because its default value is FALSE as expected.

mickaelaccensi replied:
ok, so we can create a file namelist.nml in the model/nml which contains some namelist values if you also want

ukmo-juan-castillo replied:
I can do it, adding all the possible namelist parameters with their default value.

ukmo-ansaulter replied:
Thanks Juan, the new version using CALTYPE works for this

ukmo-ansaulter reviewed (commented):
Thanks Juan,

I've added some comments in ww3_grid.ftn

Also, for your consideration:

- A CAL360 option needs to be included in the ww3_grid.nml template file (I couldn't see this, only the .inp version)?
- I'm unsure if the ww3_ts1 regtest, which is for source terms(?), is the best place to put this calendar test, or whether a dedicated new test would be better. @ukmo-ccbunney can you comment?

Cheers
Andy

mickaelaccensi commented on model/ftn/ww3_ounf.ftn:
```diff
@@ -861,7 +862,11 @@
+        CALL U2D('days since 1900-01-01 00:00:00',REFDATE,IERR)
+      ELSE
+        CALL U2D('days since 1990-01-01 00:00:00',REFDATE,IERR)
+      ENDIF
```
why do you need to set the reference date to 1900 if cal360 ?

ukmo-juan-castillo replied:
The 360-day calendar is usually employed in climatological studies, which run for several decades. If somebody wants to run a climatological study for several decades up to the present day (something that is done at the Met Office, for example), the start of the run is probably before the 1990 reference date. Of course we will have the same problem if we want to start the run before year 1900, so the choice of year is something that can be discussed. At least 1900 will allow running the most common climate jobs.

mickaelaccensi replied:
it has no impact, the value will be negative if it's before 1990. I know that some codes have still the reference date hardcoded as 1990 so it should be better to keep 1990 in all cases.

ukmo-juan-castillo replied:
In that case I will just leave 1990 as the reference date. Thanks.

ukmo-ansaulter replied:
Change looks OK to me

ukmo-ccbunney replied:
Looks good to me.

mickaelaccensi commented on model/ftn/ww3_ounp.ftn:
```diff
@@ -2665,12 +2666,25 @@
+        IRET=NF90_PUT_ATT(NCID,VARID(1),'long_name','julian day (UT)')
+      ENDIF
       IRET=NF90_PUT_ATT(NCID,VARID(1),'standard_name','time')
       IRET=NF90_PUT_ATT(NCID,VARID(1),'calendar','standard')
```
this line is duplicated with lines 2686

ukmo-juan-castillo replied:
Agree, I will remove line 2675.

ukmo-ansaulter replied:
Change looks OK to me

mickaelaccensi commented on model/ftn/ww3_ounp.ftn:
```diff
@@ -2665,12 +2666,25 @@
 !/NC4      IF (NCTYPE.EQ.4) IRET=NF90_DEF_VAR_DEFLATE(NCID, VARID(1), 1, 1, DEFLATE)
-      IRET=NF90_PUT_ATT(NCID,VARID(1),'long_name','julian day (UT)')
+      IF (CAL360) THEN
+        IRET=NF90_PUT_ATT(NCID,VARID(1),'long_name','days since 1900-01-01')
```
long_name should be described as : "All years are 360 days divided into 30 day months." regarding the CF convention 
http://cfconventions.org/Data/cf-conventions/cf-conventions-1.8/cf-conventions.html#calendar

ukmo-ccbunney replied:
@mickaelaccensi Is that correct? I think that "long_name" should actually just be "time" with an appropriate value in the "calendar" attribute.

ukmo-ansaulter replied:
@ukmo-ccbunney and @mickaelaccensi 

standard_name should be 'time'

long_name can be whatever we choose, but re Mikael's comment it is probably advisable if it indicates the form of calendar we are using in the non-standard cases (makes it an easy read for user, and the same should go for a noleap time variable?)

as Chris says, the important attribute here is 'calendar'; see http://cfconventions.org/cf-conventions/cf-conventions.html#calendar

there is always the option of the 'comment' attribute if we really aren't happy, but best if the variable can be made clear via standard_name, units, and calendar

So a  preferred answer might be something like:

long_name = 'days since 1900-01-01 in 360 day calendar'

As an aside to this change, the calendar attribute in ww3_ounf should also have a 'no_leap' or '365_day' option?

ukmo-ccbunney replied:
@ukmo-ansaulter @mickaelaccensi 
> So a preferred answer might be something like:
> long_name = 'days since 1900-01-01 in 360 day calendar'

OK, although I think in that case the units should go in the long_name. Long name is described as:

>  "_a long descriptive name which may, for example, be used for labeling plots_",

so, to my mind, it should describe the values after they have been "decoded" by having the units applied to them.
E.g. "_time in 360 day calendar_", or similar.

mickaelaccensi replied:
ok for long_name = "time in 360 day calendar"

ukmo-juan-castillo replied:
I will use the suggested long_name="time in 360 day calendar", and add long_name and calendar attributes for the no leap calendar using cf convenctions, leaving units as 'days since 1990-01-01 00:00:00' and the rest of the attributes unchanged.

ukmo-ansaulter replied:
Change looks OK to me

ukmo-ccbunney replied:
Looks good to me.

ukmo-juan-castillo committed:
Merge branch 'fb_360_calendar' of https://github.com/ukmo-waves/WW3 into fb_360_calendar

ukmo-ansaulter commented on model/ftn/ww3_grid.ftn:
```diff
@@ -6329,6 +6349,8 @@
 !/NL3 1004 FORMAT (/' *** WAVEWATCH-III ERROR IN READNL : '/          &
 !/NL3               '     TEMP QPARMS ARRAY TOO SMALL, .LE. ',I8/)
+ 1005 FORMAT (/' *** WAVEWATCH III ERROR IN READNL : '/          &
+               '     UNDETERMINED CALENDAR: CAL360 and NOLEAP'/)
```
Comment:

'     UNDETERMINED CALENDAR: Select only one of CAL360 and NOLEAP'

might be a little more explicit?

ukmo-juan-castillo replied:
Agree, will be changed

ukmo-ansaulter replied:
Happy with the new text for CALTYPE

ukmo-juan-castillo commented:
As a possible improvement, I would propose substituting the MISC namelist parameters CAL360 and NOLEAP by another one called CALENDAR, which only possible values would be 'standard' (default value), '360_day', or '365_day'. 

The changes to the code needed for this implementation are straightforward. I checked that no changes in the regtest namelists will be needed.

Finally, there is no regtest available for the 365-day calendar, and it will possibly be a good idea adding it.

Waiting for your comments before taking any action.

mickaelaccensi commented:
> As a possible improvement, I would propose substituting the MISC namelist parameters CAL360 and NOLEAP by another one called CALENDAR, which only possible values would be 'standard' (default value), '360_day', or '365_day'.
> 
> The changes to the code needed for this implementation are straightforward. I checked that no changes in the regtest namelists will be needed.
> 
> Finally, there is no regtest available for the 365-day calendar, and it will possibly be a good idea adding it.
> 
> Waiting for your comments before taking any action.

I fully agree with your idea. It will be much more clear to have a dedicated namelist instead of putting everything in MISC. also a dedicated regtest is a great idea

ukmo-juan-castillo committed:
Changes after reviewer's comments: change CAL360 and NOLEAP namelist
parameters by CALTYPE; add regtest for 365_day calendat (although the
location of this test is still to be decided)

ukmo-ccbunney reviewed (commented):
>> I'm unsure if the ww3_ts1 regtest, which is for source terms(?), is the best place to put this calendar test, or whether a dedicated new test would be better. @ukmo-ccbunney can you comment?

I wonder if a new single grid cell regtest would be more appropriate, rather than adding to ww3_ts1?

It only needs to test the dates are correct, so maybe we could utilise the "**dry run**" mode to turn off all source term integration and propagation?

ukmo-ccbunney reviewed (changes requested):
https://github.com/ukmo-waves/WW3/blob/c1890a7b803489323f9b622967be7d06e1fb13ae/model/ftn/w3iogrmd.ftn#L70

As you have modified the format of the moddef.ww3 file, this version tag will need updating.

ukmo-juan-castillo committed:
Update mod_def version and create calendar regtests

ukmo-juan-castillo commented:
The change requested regarding the version of mod_def has been addressed, and as requested new ww3_tc1 regtests have been created to test the new calendars.

Waiting for new revision.

ukmo-ccbunney reviewed (commented):
@ukmo-juan-castillo I think we should be able to put the `360day` and `365_day` calendar tests into a single input directory in ww3_tc1. You can then differentiate between them by using the `-g` flag in `run_test`. What do you think?

EDIT: Equally, using the `-i` flag works as well, but the `-g` allows us to avoid duplicating a few files. Admittedly, there is not a lot of saving.

ukmo-juan-castillo commented:
The problem is that, as far as I understand, the -g flag allows you to specify the ww3_grid.inp file, but there is no flag for a different ww3_shel.inp file (and ww3_ounf.inp), which is needed because the dates of the run are different.

ukmo-ccbunney commented:
> The problem is that, as far as I understand, the -g flag allows you to specify the ww3_grid.inp file, but there is no flag for a different ww3_shel.inp file (and ww3_ounf.inp), which is needed because the dates of the run are different.

Ah - OK, I understand the issue now.
In that case, I wonder if we should be very explicit about the naming convention and rename `input` to `input_CAL365`?

`input` could have been the test for the "default" calendar, but I assume that you consider testing the default option pointless, as all the other regtests will use the default anyhow?

ukmo-juan-castillo committed:
Simplify calendar regtests

ukmo-juan-castillo commented:
This should be the definitive version, where the regtests have been simplified so that the tests for all the possible calendars are contained in one single input directory. The results of the tests are as expected.

mickaelaccensi reviewed (approved):
thanks for this nice work and clean up! I approve

ukmo-ansaulter reviewed (approved):
Many thanks Juan

ukmo-ccbunney reviewed (approved):
Looks great. Thanks Juan.

ukmo-ccbunney commented:
@ukmo-juan-castillo - one last task before I can merge: 
Please can change the target branch from `develop` to `staging`.
Thanks.
